### PR TITLE
feat: add guild scheduled event

### PR DESCRIPTION
**Pull request by iShibi**
**Please describe the changes this PR makes and why it should be merged:**

Adds type definitions for Guild Scheduled Event

**Reference Discord API Docs PRs or commits:**
- discord/discord-api-docs#3586

### CODE REVIEW

**Comment by iShibi** [comment]
This is ready for review. Once it is released the typings error will be gone in [discordjs/discord.js#6943](https://github.com/discordjs/discord.js/pull/6493)

File: gateway/v8.ts [hunk 1]
```diff
@@ -237,6 +239,11 @@ export const enum GatewayDispatchEvents {
 	VoiceServerUpdate = 'VOICE_SERVER_UPDATE',
 	VoiceStateUpdate = 'VOICE_STATE_UPDATE',
 	WebhooksUpdate = 'WEBHOOKS_UPDATE',
+	GuildScheduledEventCreate = 'Guild_Scheduled_Event_Create',
+	GuildScheduledEventUpdate = 'Guild_Scheduled_Event_Update',
+	GuildScheduledEventDelete = 'Guild_Scheduled_Event_Delete',
+	GuildScheduledEventUserAdd = 'Guild_Scheduled_Event_User_Add',
+	GuildScheduledEventUserRemove = 'Guild_Scheduled_Event_User_Remove',
```
Comment (vladfrangu) [inline gateway/v8.ts:246]: All of these are wrong in the values, they need to be UPPER_SNAKE_CASE

File: gateway/v9.ts [hunk 2]
```diff
@@ -246,6 +248,11 @@ export const enum GatewayDispatchEvents {
 	VoiceServerUpdate = 'VOICE_SERVER_UPDATE',
 	VoiceStateUpdate = 'VOICE_STATE_UPDATE',
 	WebhooksUpdate = 'WEBHOOKS_UPDATE',
+	GuildScheduledEventCreate = 'Guild_Scheduled_Event_Create',
+	GuildScheduledEventUpdate = 'Guild_Scheduled_Event_Update',
+	GuildScheduledEventDelete = 'Guild_Scheduled_Event_Delete',
+	GuildScheduledEventUserAdd = 'Guild_Scheduled_Event_User_Add',
+	GuildScheduledEventUserRemove = 'Guild_Scheduled_Event_User_Remove',
```
Comment (vladfrangu) [inline gateway/v9.ts:255]: Ditto

File: payloads/v8/auditLog.ts [hunk 3]
```diff
@@ -611,6 +618,21 @@ export type APIAuditLogChangeKeyAvailable = AuditLogChangeData<'available', bool
  */
 export type APIAuditLogChangeKeyGuildId = AuditLogChangeData<'guild_id', Snowflake>;
 
+/**
+ * Returned when entity type of a guild scheduled event is changed
+ */
+export type APIAuditLogChangeKeyEntityType = AuditLogChangeData<'entity_type', number>;
```
Comment (vladfrangu) [inline payloads/v8/auditLog.ts:624]: GuildScheduledEventEntityType instead for the type?

File: payloads/v8/auditLog.ts [hunk 4]
```diff
@@ -611,6 +618,21 @@ export type APIAuditLogChangeKeyAvailable = AuditLogChangeData<'available', bool
...
+
+/**
+ * Returned when status of a guild scheduled event is changed
+ */
+export type APIAuditLogChangeKeyStatus = AuditLogChangeData<'status', number>;
```
Comment (vladfrangu) [inline payloads/v8/auditLog.ts:629]: GuildScheduledEventStatus instead for the type?

File: payloads/v8/guildScheduledEvent.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	 */
+	scheduled_start_time: string;
+	/**
+	 * The time at which the guild event will end, or `null` if the event does not have a scheduled time to end
+	 */
+	scheduled_end_time: string | null;
+	/**
+	 * The privacy level of the scheduled event
+	 */
+	privvacy_level: GuildScheduledEventPrivacyLevel;
```
Comment (vladfrangu) [inline payloads/v8/guildScheduledEvent.ts:40]: vvvvvv
Too many v's

File: payloads/v8/invite.ts [hunk 6]
```diff
@@ -68,6 +69,10 @@ export interface APIInvite {
 	 * The stage instance data if there is a public stage instance in the stage channel this invite is for
 	 */
 	stage_instance?: APIInviteStageInstance;
+	/**
+	 * The guild scheduled event data, only included if `guild_scheduled_event_id` contains a valid guild scheduled event id
```
Comment (vladfrangu) [inline payloads/v8/invite.ts:73]: If this is a query parameter, can you follow the same flow as the others? (like expires_at for example)

File: payloads/v9/auditLog.ts [hunk 7]
```diff
@@ -650,6 +657,21 @@ export type APIAuditLogChangeKeyDefaultAutoArchiveDuration = AuditLogChangeData<
 	number
 >;
 
+/**
+ * Returned when entity type of a guild scheduled event is changed
+ */
+export type APIAuditLogChangeKeyEntityType = AuditLogChangeData<'entity_type', number>;
```
Comment (vladfrangu) [inline payloads/v9/auditLog.ts:663]: Same as v8

File: payloads/v9/auditLog.ts [hunk 8]
```diff
@@ -650,6 +657,21 @@ export type APIAuditLogChangeKeyDefaultAutoArchiveDuration = AuditLogChangeData<
...
+
+/**
+ * Returned when status of a guild scheduled event is changed
+ */
+export type APIAuditLogChangeKeyStatus = AuditLogChangeData<'status', number>;
+
+/**
+ * Returned when location of a guild scheduled event is changed
+ */
```
Comment (vladfrangu) [inline payloads/v9/auditLog.ts:672]: Same as v8

File: payloads/v9/guildScheduledEvent.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	 */
+	scheduled_start_time: string;
+	/**
+	 * The time at which the guild event will end, or `null` if the event does not have a scheduled time to end
+	 */
+	scheduled_end_time: string | null;
+	/**
+	 * The privacy level of the scheduled event
+	 */
+	privvacy_level: GuildScheduledEventPrivacyLevel;
```
Comment (vladfrangu) [inline payloads/v9/guildScheduledEvent.ts:40]: Same as v8

File: payloads/v9/invite.ts [hunk 10]
```diff
@@ -68,6 +69,10 @@ export interface APIInvite {
 	 * The stage instance data if there is a public stage instance in the stage channel this invite is for
 	 */
 	stage_instance?: APIInviteStageInstance;
+	/**
+	 * The guild scheduled event data, only included if `guild_scheduled_event_id` contains a valid guild scheduled event id
```
Comment (vladfrangu) [inline payloads/v9/invite.ts:73]: Same as v8

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: payloads/v8/guildScheduledEvent.ts (hunk 5 above)
Comment (iShibi) [reply payloads/v8/guildScheduledEvent.ts]: ✌️ 😄

File: payloads/v9/auditLog.ts (hunk 8 above)
Comment (iShibi) [reply payloads/v9/auditLog.ts:673]: I can't find the v8 version of this suggestion, was this a mistake?

**Comment by iShibi** [comment]
Added new upstream changes:
- https://github.com/discord/discord-api-docs/pull/4113

File: payloads/v9/auditLog.ts (hunk 8 above)
Comment (vladfrangu) [reply payloads/v9/auditLog.ts:673]: I probably replied to the wrong line then, basically what I reported in v9 should also be fixed in v8

**Comment by iShibi** [comment]
Added some more changes:
- https://github.com/discord/discord-api-docs/pull/3919
- https://github.com/discord/discord-api-docs/pull/4117

File: rest/common.ts [hunk 11]
```diff
@@ -102,6 +102,8 @@ export const enum RESTJSONErrorCodes {
 
 	MaximumNumberOfBanFetchesHasBeenReached = 30037,
 
+	MaximumNumberOfUncompletedGuildScheduledEventsReached = 30038,
```
Comment (vladfrangu) [inline rest/common.ts:105]:
```suggestion
	MaximumNumberOfBanFetchesHasBeenReached = 30037,
	MaximumNumberOfUncompletedGuildScheduledEventsReached,
```

**Review by vladfrangu — CHANGES_REQUESTED** [review]
Small nit, if you can apply this for deno as well, we can get this merged

**Review by vladfrangu — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Comment by vladfrangu** [comment]
@iShibi can you confirm that this is everything, so I can get this merged? If there are changes after merge, feel free to PR them :D

**Comment by iShibi** [comment]
@vladfrangu Yup, everything that has been merged is here. The only thing that's left is this:

- https://github.com/discord/discord-api-docs/pull/4152

**Comment by vladfrangu** [comment]
I'm going to merge this as is to finally actually release 0.25.0 (a month of changes...wooo) so you'll have to PR audit log stuffs when they merge. Worst case I'll do a quick 0.25.1 for that

**Pull request merged**